File: Desktop/tableau-ai-analysis-assistant-main/tableau_assistant/src/api/streaming.py
"""
流式API端点

提供SSE(Server-Sent Events)接口用于实时进度反馈
"""
from fastapi import APIRouter, Request
from fastapi.responses import StreamingResponse
from typing import Dict, Any
import json

from tableau_assistant.tests.test_workflow import create_test_workflow
from tableau_assistant.src.workflows.streaming import stream_workflow_events

router = APIRouter(prefix="/api/stream", tags=["streaming"])

# 用于测试的工作流创建函数（可以被测试覆盖）
_workflow_factory = create_test_workflow


@router.post("/chat")
async def stream_chat(request: Request):
    """
    流式对话API
    
    使用SSE推送实时进度
    
    Request Body:
        {
            "question": str,  # 用户问题
            "datasource_luid": str,  # 数据源LUID
            "user_id": str,  # 用户ID
            "session_id": str,  # 会话ID
            "boost_question": bool  # 是否使用问题Boost（可选）
        }
    
    Response:
        SSE流，事件格式:
        {
            "type": str,  # 事件类型
            "data": dict,  # 事件数据
            "timestamp": float  # 时间戳
        }
    """
    # 解析请求
    body = await request.json()
    
    question = body.get("question")
    datasource_luid = body.get("datasource_luid")
    user_id = body.get("user_id")
    session_id = body.get("session_id", user_id)
    boost_question = body.get("boost_question", False)
    
    # 验证必需参数
    if not question:
        return {"error": "question is required"}
    if not datasource_luid:
        return {"error": "datasource_luid is required"}
    if not user_id:
        return {"error": "user_id is required"}
    
    # 创建工作流
    app = _workflow_factory()
    
    # 准备输入
    input_data = {
        "question": question,
        "boost_question": boost_question
    }
    
    # 准备config
    config = {
        "configurable": {
            "thread_id": session_id,
            "datasource_luid": datasource_luid,
            "user_id": user_id,
            "session_id": session_id,
            "max_replan_rounds": 2,
            "parallel_upper_limit": 3,
            "max_retry_times": 2,
            "max_subtasks_per_round": 10
        }
    }
    
    # 返回SSE流
    return StreamingResponse(
        stream_workflow_events(app, input_data, config),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no"  # 禁用nginx缓冲
        }
    )


@router.get("/health")
async def health_check():
    """健康检查"""
    return {"status": "ok"}


# 示例：如何在前端使用SSE
"""
// JavaScript示例
const eventSource = new EventSource('/api/stream/chat', {
    method: 'POST',
    headers: {
        'Content-Type': 'application/json'
    },
    body: JSON.stringify({
        question: '2016年各地区的销售额',
        datasource_luid: 'abc123',
        user_id: 'user_456'
    })
});

eventSource.onmessage = (event) => {
    const data = JSON.parse(event.data);
    console.log('Event:', data.type, data.data);
    
    switch (data.type) {
        case 'token':
            // Token级流式渲染
            appendToken(data.data.token);
            break;
        case 'agent_start':
            // 显示Agent开始执行
            showAgentProgress(data.data.agent, 'running');
            break;
        case 'agent_complete':
            // 显示Agent完成
            showAgentProgress(data.data.agent, 'complete');
            break;
        case 'workflow_complete':
            // 工作流完成
            eventSource.close();
            break;
    }
};

eventSource.onerror = (error) => {
    console.error('SSE Error:', error);
    eventSource.close();
};
"""

File: Desktop/tableau-ai-analysis-assistant-main/tableau_extension/src/api/streaming.ts
/**
 * SSE 流式客户端
 * 
 * 用于接收后端的 Server-Sent Events (SSE) 流式数据
 */

// @ts-ignore - Vite env variables
const API_BASE_URL = import.meta.env?.VITE_API_BASE_URL

/**
 * 事件类型枚举
 */
export enum EventType {
  // Token级流式输出
  TOKEN = 'token',
  
  // Agent进度
  AGENT_START = 'agent_start',
  AGENT_COMPLETE = 'agent_complete',
  AGENT_ERROR = 'agent_error',
  
  // 工具调用进度
  TOOL_START = 'tool_start',
  TOOL_COMPLETE = 'tool_complete',
  TOOL_ERROR = 'tool_error',
  
  // 查询执行进度
  QUERY_START = 'query_start',
  QUERY_COMPLETE = 'query_complete',
  QUERY_ERROR = 'query_error',
  
  // 整体进度
  WORKFLOW_START = 'workflow_start',
  WORKFLOW_COMPLETE = 'workflow_complete',
  WORKFLOW_ERROR = 'workflow_error',
  
  // 其他
  PROGRESS = 'progress',
  LOG = 'log'
}

/**
 * 流式事件接口
 */
export interface StreamEvent {
  type: EventType
  data: any
  timestamp: number
}

/**
 * Token事件数据
 */
export interface TokenEventData {
  token: string
  agent?: string
}

/**
 * Agent事件数据
 */
export interface AgentEventData {
  agent: string
  run_id?: string
  output?: any
  error?: string
  timestamp: number
}

/**
 * 工具事件数据
 */
export interface ToolEventData {
  tool: string
  input?: any
  output?: any
  error?: string
  timestamp: number
}

/**
 * 工作流事件数据
 */
export interface WorkflowEventData {
  question?: string
  duration?: number
  error?: string
  timestamp: number
}

/**
 * 事件处理器类型
 */
export type EventHandler = (event: StreamEvent) => void

/**
 * 聊天请求参数
 */
export interface ChatRequest {
  question: string
  datasource_luid: string
  user_id: string
  session_id?: string
  boost_question?: boolean
}

/**
 * SSE 流式客户端类
 */
export class StreamingClient {
  private eventSource: EventSource | null = null
  private handlers: Map<EventType | 'all', EventHandler[]> = new Map()

  /**
   * 开始流式聊天
   */
  async startChat(request: ChatRequest): Promise<void> {
    // 关闭现有连接
    this.close()

    // 创建 EventSource
    // 注意：EventSource 不支持 POST，所以我们需要使用 fetch + ReadableStream
    await this.streamWithFetch(request)
  }

  /**
   * 使用 fetch API 实现 SSE 流式请求
   */
  private async streamWithFetch(request: ChatRequest): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/api/stream/chat`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(request)
      })

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`)
      }

      const reader = response.body?.getReader()
      if (!reader) {
        throw new Error('Response body is not readable')
      }

      const decoder = new TextDecoder()
      let buffer = ''

      while (true) {
        const { done, value } = await reader.read()
        
        if (done) {
          break
        }

        // 解码数据
        buffer += decoder.decode(value, { stream: true })

        // 处理完整的 SSE 消息
        const lines = buffer.split('\n\n')
        buffer = lines.pop() || '' // 保留不完整的消息

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            const jsonStr = line.substring(6) // 去掉 "data: "
            try {
              const event: StreamEvent = JSON.parse(jsonStr)
              this.handleEvent(event)
            } catch (error) {
              console.error('Failed to parse SSE event:', error, jsonStr)
            }
          }
        }
      }
    } catch (error) {
      console.error('Streaming error:', error)
      // 触发错误事件
      this.handleEvent({
        type: EventType.WORKFLOW_ERROR,
        data: { error: String(error) },
        timestamp: Date.now() / 1000
      })
    }
  }

  /**
   * 注册事件处理器
   */
  on(eventType: EventType | 'all', handler: EventHandler): void {
    if (!this.handlers.has(eventType)) {
      this.handlers.set(eventType, [])
    }
    this.handlers.get(eventType)!.push(handler)
  }

  /**
   * 移除事件处理器
   */
  off(eventType: EventType | 'all', handler: EventHandler): void {
    const handlers = this.handlers.get(eventType)
    if (handlers) {
      const index = handlers.indexOf(handler)
      if (index > -1) {
        handlers.splice(index, 1)
      }
    }
  }

  /**
   * 处理事件
   */
  private handleEvent(event: StreamEvent): void {
    // 触发特定类型的处理器
    const typeHandlers = this.handlers.get(event.type as EventType)
    if (typeHandlers) {
      typeHandlers.forEach(handler => handler(event))
    }

    // 触发通用处理器
    const allHandlers = this.handlers.get('all')
    if (allHandlers) {
      allHandlers.forEach(handler => handler(event))
    }
  }

  /**
   * 关闭连接
   */
  close(): void {
    if (this.eventSource) {
      this.eventSource.close()
      this.eventSource = null
    }
  }

  /**
   * 清除所有事件处理器
   */
  clearHandlers(): void {
    this.handlers.clear()
  }
}

/**
 * 创建流式客户端实例
 */
export function createStreamingClient(): StreamingClient {
  return new StreamingClient()
}
